tfhe/src/integer/server_key/radix_parallel/sub.rs

```diff
@@ -402,6 +413,329 @@ impl ServerKey {
+
+                r << (i - 1)
+            };
+            first_grouping_luts.push(
+                self.key
+                    .generate_many_lookup_table(&[&state_fn, &shift_block_fn]),
+            );
+        }
+
+        let other_block_state_luts = (0..grouping_size)
+            .map(|i| {
+                let state_fn = |block| {
+                    #[allow(clippy::comparison_chain)]
+                    let r = if block < message_modulus {
+                        2 // Generates borrow
+                    } else if block == message_modulus {
+                        1 // Propagates a carry
+                    } else {
+                        0 // Does not borrow
+                    };
+
+                    r << i
+                };
+                self.key
+                    .generate_many_lookup_table(&[&state_fn, &shift_block_fn])
+            })
+            .collect::<Vec<_>>();
+
+        // For the last block we do something a bit different because the
+        // state we compute will be used (if needed) to compute the output borrow
+        // of the whole subtraction. And this computation will be done during the 'cleaning'
+        // phase
+        let last_block_luts = {
+            if blocks.len() == 1 {
+                let first_block_state_fn = |block| {
+                    if block < message_modulus {
+                        2 << 1 // Generates a borrow
+                    } else {
+                        0 // Nothing
+                    }
+                };
+                self.key
+                    .generate_many_lookup_table(&[&first_block_state_fn, &shift_block_fn])
+            } else {
+                first_grouping_luts[2].clone()
+            }
+        };
+
+        let tmp = blocks
+            .par_iter()
+            .enumerate()
+            .map(|(index, block)| {
+                let grouping_index = index / grouping_size;
+                let is_in_first_grouping = grouping_index == 0;
+                let index_in_grouping = index % (grouping_size);
+                let is_last_index = index == blocks.len() - 1;
+
+                let luts = if is_last_index {
+                    &last_block_luts
+                } else if is_in_first_grouping {
+                    &first_grouping_luts[index_in_grouping]
+                } else {
+                    &other_block_state_luts[index_in_grouping]
+                };
+                self.key.apply_many_lookup_table(block, luts)
+            })
+            .collect::<Vec<_>>();
+
+        let mut shifted_blocks = Vec::with_capacity(num_blocks);
+        let mut block_states = Vec::with_capacity(num_blocks);
+        for mut blocks in tmp {
+            assert_eq!(blocks.len(), 2);
+            shifted_blocks.push(blocks.pop().unwrap());
+            block_states.push(blocks.pop().unwrap());
+        }
+
+        (shifted_blocks, block_states)
+    }
+
+    pub(crate) fn advanced_sub_assign_with_borrow_sequential(
```

Comment:
obligatory: this function has its own dedicated somehow to make sure we are testing it ?